ghcide/src/Development/IDE/Core/Shake.hs

```diff
@@ -1088,24 +1105,26 @@ defineEarlyCutoff' doDiagnostics cmp key file old mode action = do
         res <- case val of
             Just res -> return res
             Nothing -> do
+                staleV <- liftIO $ atomicallyNamed "define -read 3" $ getValues state key file <&> \case
+                  Nothing -> Failed False
+                  Just (Succeeded ver v, _) -> Stale Nothing ver v
+                  Just (Stale d ver v, _)   -> Stale d ver v
+                  Just (Failed b, _)        -> Failed b
```

Comment:
We don't need to compute `staleV` when the rule runs successfully, right?

Reply:
True, but together with https://github.com/haskell/haskell-language-server/pull/2316/commits/3273b87205b011f7acccc1c7a534d72d481054f1 in #2316, we will need to compute it in all cases. We can close this PR and merge as a batch with #2316 if that is any better?

Reply:
I don't see any problem with changing it again in #2316